blob/src/main/java/com/bazaarvoice/emodb/blob/db/s3/S3StorageProvider.java

```diff
@@ -0,0 +1,435 @@
+import com.amazonaws.services.s3.model.InitiateMultipartUploadResult;
+import com.amazonaws.services.s3.model.ListObjectsV2Request;
+import com.amazonaws.services.s3.model.ListObjectsV2Result;
+import com.amazonaws.services.s3.model.ObjectMetadata;
+import com.amazonaws.services.s3.model.PartETag;
+import com.amazonaws.services.s3.model.PutObjectRequest;
+import com.amazonaws.services.s3.model.S3Object;
+import com.amazonaws.services.s3.model.S3ObjectSummary;
+import com.amazonaws.services.s3.model.UploadPartRequest;
+import com.amazonaws.services.s3.model.UploadPartResult;
+import com.amazonaws.util.IOUtils;
+import com.bazaarvoice.emodb.blob.api.Range;
+import com.bazaarvoice.emodb.blob.api.StreamSupplier;
+import com.bazaarvoice.emodb.blob.core.BlobPlacement;
+import com.bazaarvoice.emodb.table.db.astyanax.PlacementFactory;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import com.google.common.io.ByteStreams;
+import com.netflix.astyanax.connectionpool.exceptions.ConnectionException;
+import org.apache.commons.codec.binary.Hex;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import javax.inject.Inject;
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.nio.ByteBuffer;
+import java.security.DigestInputStream;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Date;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public class S3StorageProvider {
+    private static final Logger LOGGER = LoggerFactory.getLogger(S3StorageProvider.class);
+
+    private final PlacementFactory _placementFactory;
+    private final LoadingCache<String, BlobPlacement> _placementCache;
+
+    @Inject
+    public S3StorageProvider(final PlacementFactory placementFactory) {
+        _placementFactory = Objects.requireNonNull(placementFactory);
+        _placementCache = CacheBuilder.newBuilder().build(new CacheLoader<String, BlobPlacement>() {
+            @Override
+            public BlobPlacement load(@Nonnull final String placement) throws ConnectionException {
+                BlobPlacement blobPlacement = (BlobPlacement) _placementFactory.newPlacement(placement);
+                LOGGER.debug("Blob placement: {}", blobPlacement);
+                return blobPlacement;
+            }
+        });
+    }
+
+    private static AmazonS3URI getAmazonS3URI(final String bucket, final String tableName, @Nullable final String blobId) {
```

Comment:
Even though it is convenient to make the s3 uri essentially {table}/{key}, I don't think this can be safe in the face of a deleted table. Imagine if we created a table, wrote some blobs, deleted the table, then recreated it again. We would have no way of differentiating the original table from the new one sharing the same name. The best way to fix this is to use Emo's Row Key logic it uses in Cassandra in S3. Let's delve deeper into this offline after standup this week.

Reply:
agreed, a real problem!